buffer/src/main/java/io/netty/buffer/api/Buffer.java

```diff
@@ -0,0 +1,685 @@
+     * {@link #readerOffset()} or greater than {@link #capacity()}.
+     * @throws BufferClosedException if this buffer is closed.
+     * @throws BufferReadOnlyException if this buffer is {@linkplain #readOnly() read-only}.
+     */
+    Buffer writerOffset(int offset);
+
+    /**
+     * Returns the number of readable bytes which is equal to {@code (writerOffset() - readerOffset())}.
+     */
+    default int readableBytes() {
+        return writerOffset() - readerOffset();
+    }
+
+    /**
+     * Returns the number of writable bytes which is equal to {@code (capacity() - writerOffset())}.
+     */
+    default int writableBytes() {
+        return capacity() - writerOffset();
+    }
+
+    /**
+     * Fills the buffer with the given byte value. This method does not respect the {@link #readerOffset()} or {@link
+     * #writerOffset()}, but copies the full capacity of the buffer. The {@link #readerOffset()} and {@link
+     * #writerOffset()} are not modified.
+     *
+     * @param value The byte value to write at every position in the buffer.
+     * @return This Buffer.
+     * @throws BufferReadOnlyException if this buffer is {@linkplain #readOnly() read-only}.
+     */
+    Buffer fill(byte value);
+
+    /**
+     * Gives the native memory address backing this buffer, or return 0 if this buffer has no native memory address.
+     * @return The native memory address, if any, otherwise 0.
+     */
+    long nativeAddress();
+
+    /**
+     * Makes this buffer read-only.
+     * This is irreversible.
+     *
+     * @return this buffer.
+     */
+    Buffer makeReadOnly();
+
+    /**
+     * Queries if this buffer is read-only or not.
+     *
+     * @return {@code true} if this buffer is read-only, {@code false} otherwise.
+     */
+    boolean readOnly();
+
+    /**
+     * Copies the given length of data from this buffer into the given destination array, beginning at the given source
+     * position in this buffer, and the given destination position in the destination array.
+     * <p>
+     * This method does not read or modify the {@linkplain #writerOffset() write offset} or the
+     * {@linkplain #readerOffset() read offset}.
+     *
+     * @param srcPos The byte offset into this buffer wherefrom the copying should start; the byte at this offset in
+     *              this buffer will be copied to the {@code destPos} index in the {@code dest} array.
+     * @param dest The destination byte array.
+     * @param destPos The index into the {@code dest} array wherefrom the copying should start.
+     * @param length The number of bytes to copy.
+     * @throws NullPointerException if the destination array is null.
+     * @throws IndexOutOfBoundsException if the source or destination positions, or the length, are negative,
+     * or if the resulting end positions reaches beyond the end of either this buffer, or the destination array.
+     */
+    void copyInto(int srcPos, byte[] dest, int destPos, int length);
```

Comment:
I see your point, but I have to disagree, because it will bring `srcPos` and `destPos` next to each other, and that can make it unclear which is which at a call site. The `ByteBuf.getBytes` family of methods have the same argument layout.